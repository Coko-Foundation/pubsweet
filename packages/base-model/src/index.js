const uuid = require('uuid')
const { Model, ValidationError, transaction } = require('objection')
const logger = require('@pubsweet/logger')
const { db, NotFoundError } = require('pubsweet-server')
const { merge } = require('lodash')
const config = require('config')

Model.knex(db)

const validationError = (prop, className) =>
  new ValidationError({
    type: 'ModelValidation',
    message: `${prop} is not a property in ${className}'s schema`,
  })

const notFoundError = (property, value, className) =>
  new NotFoundError(`Object not found: ${className} with ${property} ${value}`)

const integrityError = (property, value, message) =>
  new Error(
    `Data Integrity Error property ${property} set to ${value}: ${message}`,
  )

class BaseModel extends Model {
  constructor(properties) {
    super(properties)

    if (properties) {
      this.updateProperties(properties)
    }

    const handler = {
      set: (obj, prop, value) => {
        if (this.isSettable(prop)) {
          obj[prop] = value
          return true
        }

        throw validationError(prop, obj.constructor.name)
      },
    }

    return new Proxy(this, handler)
  }

  static get jsonSchema() {
    let schema

    const mergeSchema = additionalSchema => {
      if (additionalSchema) {
        schema = merge(schema, additionalSchema)
      }
    }

    // Crawls up the prototype chain to collect schema
    // information from models and extended models
    const getSchemasRecursively = object => {
      mergeSchema(object.schema)
      mergeSchema(config.schema[object.name])

      const proto = Object.getPrototypeOf(object)

      if (proto.name !== 'BaseModel') {
        getSchemasRecursively(proto)
      }
    }

    getSchemasRecursively(this)

    const baseSchema = {
      type: 'object',
      properties: {
        type: { type: 'string' },
        id: { type: 'string', format: 'uuid' },
        created: { type: ['string', 'object'], format: 'date-time' },
        updated: {
          anyOf: [
            { type: ['string', 'object'], format: 'date-time' },
            { type: 'null' },
          ],
        },
      },
      additionalProperties: false,
    }

    if (schema) {
      return merge(baseSchema, schema)
    }
    return baseSchema
  }

  isSettable(prop) {
    const special = ['#id', '#ref']
    return (
      special.includes(prop) ||
      this.constructor.jsonSchema.properties[prop] ||
      (this.constructor.relationMappings &&
        this.constructor.relationMappings[prop])
    )
  }

  $beforeInsert() {
    this.id = this.id || uuid.v4()
    this.created = new Date().toISOString()
    this.updated = this.created
  }

  $beforeUpdate() {
    this.updated = new Date().toISOString()
  }

  async save() {
    const simpleSave = async (trx = null) =>
      this.constructor.query(trx).patchAndFetchById(this.id, this.toJSON())

    const protectedSave = async () => {
      let trx, saved
      try {
        trx = await transaction.start(BaseModel.knex())
        const current = await this.constructor.query(trx).findById(this.id)

        const storedUpdateTime = new Date(current.updated).getTime()
        const instanceUpdateTime = new Date(this.updated).getTime()

        if (instanceUpdateTime < storedUpdateTime) {
          throw integrityError(
            'updated',
            this.updated,
            'is older than the one stored in the database!',
          )
        }

        saved = await simpleSave(trx)
        await trx.commit()
      } catch (err) {
        logger.error(err)
        await trx.rollback()
        throw err
      }
      return saved
    }

    // start of save function...

    let saved
    if (this.id) {
      if (!this.updated && this.created) {
        throw integrityError(
          'updated',
          this.updated,
          'must be set when created is set!',
        )
      }

      if (!this.updated && !this.created) {
        saved = await simpleSave()
      } else {
        saved = await protectedSave()
      }
    }
    if (!saved) {
      // either model has no ID or the ID was not found in the database
      saved = await this.constructor.query().insert(this.toJSON())
    }
    logger.info(`Saved ${this.constructor.name} with UUID ${saved.id}`)
    return saved
  }

  async delete() {
    const deleted = await this.$query().delete()
    logger.info(`Deleted ${deleted} ${this.constructor.name} records.`)
    return this
  }

  updateProperties(properties) {
    Object.keys(properties).forEach(prop => {
      if (this.isSettable(prop)) {
        this[prop] = properties[prop]
      } else {
        throw validationError(prop, this.constructor.name)
      }
    })
  }

  setOwners(owners) {
    this.owners = owners
  }

  static async find(id) {
    const object = await this.query().findById(id)

    if (!object) {
      throw notFoundError('id', id, this.constructor.name)
    }

    return object
  }

  static async findByField(field, value) {
    logger.debug('Finding', field, value)

    const results = await this.query().where(field, value)

    return results
  }

  static async findOneByField(field, value) {
    const results = await this.query()
      .where(field, value)
      .limit(1)
    if (!results.length) {
      throw notFoundError(field, value, this.constructor.name)
    }

    return results[0]
  }

  static async all() {
    return this.query()
  }
}

BaseModel.pickJsonSchemaProperties = true
module.exports = BaseModel
