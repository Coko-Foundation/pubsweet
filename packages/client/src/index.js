export { default as configureStore } from './store/configureStore'
export { default as Root } from './components/Root'
export {
  default as AuthenticatedComponent,
} from './components/AuthenticatedComponent'
export { default as actions } from './actions'
export { default as reducers } from './reducers'
export { default as validations } from './validations'
export { default as withLoader } from './helpers/withLoader'
export { default as withAuthsome } from './helpers/withAuthsome'
export { default as withMutationState } from './helpers/withMutationState'
