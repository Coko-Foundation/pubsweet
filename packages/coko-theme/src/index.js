/* eslint-disable import/extensions */
import 'typeface-fira-sans-condensed'
import 'typeface-vollkorn'

import { headingScale } from '@pubsweet/ui-toolkit'

import { injectGlobal } from 'styled-components'

import { Action, ActionGroup, AppBar, Button, TextField } from './elements'

const fontSizeBase = 16
const scale = 1.2
const gridUnit = 24

injectGlobal`
  body {
    height: 100vh;
  }

  #root, #root > div {
    height: 100%;
  }
`

const cokoTheme = {
  /* Colors */
  colorBackground: 'white',
  colorPrimary: '#0B65CB',
  colorSecondary: '#E7E7E7',
  colorFurniture: '#CCC',
  colorBorder: '#AAA',
  colorBackgroundHue: '#F1F1F1',
  colorSuccess: '#008800',
  colorError: '#FF2D1A',
  colorText: '#111',
  colorTextReverse: '#FFF',
  colorTextPlaceholder: '#595959',
  colorWarning: '#ffc107',

  // TODO -- not used anywhere
  //   $colorInteract: var($colorPrimaryDarker);
  // $colorPrimaryDarker: #0551a8

  /* Text variables */
  fontInterface: "'Fira Sans Condensed'",
  fontHeading: "'Fira Sans Condensed'",
  fontReading: "'Vollkorn'",
  fontWriting: "'Cokourier Prime Sans'",
  fontSizeBase: `${fontSizeBase}px`,
  fontSizeBaseSmall: '14px',
  fontSizeHeading1: `${headingScale(fontSizeBase, scale, 1)}px`,
  fontSizeHeading2: `${headingScale(fontSizeBase, scale, 2)}px`,
  fontSizeHeading3: `${headingScale(fontSizeBase, scale, 3)}px`,
  fontSizeHeading4: `${headingScale(fontSizeBase, scale, 4)}px`,
  fontSizeHeading5: `${headingScale(fontSizeBase, scale, 5)}px`,
  fontSizeHeading6: `${headingScale(fontSizeBase, scale, 6)}px`,
  fontLineHeight: '24px',

  /* Spacing */
  gridUnit: `${gridUnit}px`,
  subGridUnit: '6px',

  /* Border */
  borderRadius: '0',
  borderWidth: '1px', // julien: not 0
  borderStyle: 'solid',

  // Does not exist
  // $borderColor: var($colorFurniture);

  /* Shadow (for tooltip) */
  boxShadow: '0 2px 4px 0 rgba(51, 51, 51, 0.3)',

  /* Transition */
  transitionDuration: '0.2s', // TODO -- julien: not 0.05s
  transitionTimingFunction: 'ease',
  transitionDelay: '0',

  cssOverrides: {
    ui: {
      Action,
      ActionGroup,
      AppBar,
      Button,
      TextField,
    },
  },
}

// console.log(cokoTheme)

export default cokoTheme
