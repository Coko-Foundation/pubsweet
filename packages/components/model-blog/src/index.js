module.exports = {
  ...require('./graphql'),
  modelName: 'Collection',
  model: require('./blog'),
  extending: '@pubsweet/model-collection',
}
