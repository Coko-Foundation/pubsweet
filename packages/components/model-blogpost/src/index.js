module.exports = {
  ...require('./graphql'),
  modelName: 'Fragment',
  model: require('./blogpost'),
  extending: '@pubsweet/model-fragment',
}
