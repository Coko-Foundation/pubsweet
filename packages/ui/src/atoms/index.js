export { default as Colorize } from './Colorize'
export { default as UploadingFile } from './UploadingFile'
export { default as Icon } from './Icon'
export { default as Attachment } from './Attachment'
export { default as Avatar } from './Avatar'
export { default as Badge } from './Badge'
export { default as Button } from './Button'
export { default as Checkbox } from './Checkbox'
export { default as File } from './File'
export { default as Menu } from './Menu'
export { default as Radio } from './Radio'
export { default as Tags } from './Tags'
export { default as Spinner } from './Spinner'
export { default as TextField } from './TextField'
export { default as ValidatedField } from './ValidatedField'
export { default as StateItem } from './StateItem'
export { default as AlignmentBox } from './AlignmentBox'
export { default as UploadButton } from './UploadButton'
export { default as Flexbox } from './Flexbox'
export { default as Link } from './Link'
export { default as Heading, H1, H2, H3, H4, H5, H6 } from './Heading'
export { default as ErrorText } from './ErrorText'
export { default as CenteredColumn } from './CenteredColumn'
export { default as Section } from './Section'
export { default as FilePicker } from './FilePicker'
