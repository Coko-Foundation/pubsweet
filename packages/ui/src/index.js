/* atoms */
export { default as Attachment } from './atoms/Attachment'
export { default as Avatar } from './atoms/Avatar'
export { default as Badge } from './atoms/Badge'
export { default as Button } from './atoms/Button'
export { default as Checkbox } from './atoms/Checkbox'
export { default as File } from './atoms/File'
export { default as Icon } from './atoms/Icon'
export { default as Menu } from './atoms/Menu'
export { default as Radio } from './atoms/Radio'
export { default as Tags } from './atoms/Tags'
export { default as TextField } from './atoms/TextField'
export { default as ValidatedField } from './atoms/ValidatedField'
export { default as StateItem } from './atoms/StateItem'
export { default as AlignmentBox } from './atoms/AlignmentBox'
export { default as UploadButton } from './atoms/UploadButton'
export { default as ControlGroup } from './atoms/ControlGroup'
export { default as Link } from './atoms/Link'
export { default as Title } from './atoms/Title'
export { default as ErrorText } from './atoms/ErrorText'
export { default as CenteredColumn } from './atoms/CenteredColumn'

/* molecules */
export { default as AppBar } from './molecules/AppBar'
export { default as Attachments } from './molecules/Attachments'
export { default as CheckboxGroup } from './molecules/CheckboxGroup'
export { default as EditableValue } from './molecules/EditableValue'
export { default as Files } from './molecules/Files'
export { default as FileUploadList } from './molecules/FileUploadList'
export { default as PlainButton } from './molecules/PlainButton'
export { default as Supplementary } from './molecules/Supplementary'
export { default as RadioGroup } from './molecules/RadioGroup'
export { default as YesOrNo } from './molecules/YesOrNo'
export { default as StateList } from './molecules/StateList'
export {
  default as AlignmentBoxWithLabel,
} from './molecules/AlignmentBoxWithLabel'
export { default as AlignmentTool } from './molecules/AlignmentTool.js'
export { default as Steps } from './molecules/Steps'

/* helpers */
export { default as theme } from './helpers/theme'
