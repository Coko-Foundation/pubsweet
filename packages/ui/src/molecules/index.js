export { default as Accordion } from './Accordion'
export { default as Action } from './Action'
export { default as ActionGroup } from './ActionGroup'
export { default as AlignmentBoxWithLabel } from './AlignmentBoxWithLabel'
export { default as AlignmentTool } from './AlignmentTool'
export { default as AppBar } from './AppBar'
export { default as Attachments } from './Attachments'
export { default as CheckboxGroup } from './CheckboxGroup'
export { default as EditableValue } from './EditableValue'
export { default as Files } from './Files'
export { default as FileUploadList } from './FileUploadList'
export { default as Menu } from './Menu'
export { default as PlainButton } from './PlainButton'
export { default as RadioGroup } from './RadioGroup'
export { default as StateList } from './StateList'
export { default as Steps } from './Steps'
export { default as Supplementary } from './Supplementary'
export { default as Upload } from './Upload'
export { default as YesOrNo } from './YesOrNo'
export { default as Tabs } from './Tabs'
