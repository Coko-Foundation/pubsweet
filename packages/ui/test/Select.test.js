import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import TestThemeProvider from './setup/theme'

import { Select } from '../src'

describe('Select', () => {
  const options = [
    { value: 'first', label: 'First' },
    { value: 'second', label: 'Second' },
    { value: 'third', label: 'Third' },
    { value: 'fourth', label: 'Fourth' },
  ]
  test('renders correctly', () => {
    const tree = renderer
      .create(
        <TestThemeProvider>
          <Select options={options} />
        </TestThemeProvider>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
