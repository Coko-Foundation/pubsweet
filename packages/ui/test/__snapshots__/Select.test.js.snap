// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly 1`] = `
<div
  className=" css-3x5r4n-container"
  onKeyDown={[Function]}
>
  <div
    className="react-select__control css-44g3o0-Control"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="react-select__value-container css-86rg1c-ValueContainer"
    >
      <div
        className="react-select__placeholder css-1i2t3f2-placeholder"
      >
        Select...
      </div>
      <div
        className="css-f7me8t-Input"
      >
        <div
          className="react-select__input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="list"
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "boxSizing": "content-box",
                "color": "inherit",
                "fontSize": "inherit",
                "label": "input",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "1px",
              }
            }
            tabIndex="0"
            type="text"
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
    </div>
    <div
      className="react-select__indicators css-1hb7zxy-IndicatorsContainer"
    >
      <span
        className="react-select__indicator-separator css-43ykx9-indicatorSeparator"
      />
      <div
        aria-hidden="true"
        className="react-select__indicator react-select__dropdown-indicator css-19njzhr-indicatorContainer"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <svg
          aria-hidden="true"
          className="css-6q0nyr-Svg"
          focusable="false"
          height={20}
          viewBox="0 0 20 20"
          width={20}
        >
          <path
            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select should render my component 1`] = `
<ThemeProvider
  theme={
    Object {
      "borderRadius": "8px",
      "borderStyle": "solid",
      "borderWidth": "1px",
      "boxShadow": "none",
      "colorBackground": "white",
      "colorBackgroundHue": "grey",
      "colorBorder": "grey",
      "colorError": "red",
      "colorFurniture": "grey",
      "colorPrimary": "black",
      "colorSecondary": "grey",
      "colorSuccess": "green",
      "colorText": "black",
      "colorTextPlaceholder": "grey",
      "colorTextReverse": "white",
      "fontHeading": "serif",
      "fontInterface": "sans-serif",
      "fontReading": "serif",
      "fontSizeBase": "18px",
      "fontSizeBaseSmall": "16px",
      "fontSizeHeading1": "36px",
      "fontSizeHeading2": "32px",
      "fontSizeHeading3": "29px",
      "fontSizeHeading4": "26px",
      "fontSizeHeading5": "23px",
      "fontSizeHeading6": "20px",
      "fontWriting": "monospace",
      "gridUnit": "8px",
      "lineHeightBase": "24px",
      "lineHeightBaseSmall": "24px",
      "lineHeightHeading1": "40px",
      "lineHeightHeading2": "40px",
      "lineHeightHeading3": "32px",
      "lineHeightHeading4": "32px",
      "lineHeightHeading5": "24px",
      "lineHeightHeading6": "24px",
      "transitionDelay": "500ms",
      "transitionDuration": "1s",
      "transitionDurationM": "0.5s",
      "transitionDurationS": "0.2s",
      "transitionDurationXs": "0.1s",
      "transitionTimingFunction": "ease",
    }
  }
>
  <WithTheme(StyledSelect)
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "First",
          "value": "first",
        },
        Object {
          "label": "Second",
          "value": "second",
        },
        Object {
          "label": "Third",
          "value": "third",
        },
        Object {
          "label": "Fourth",
          "value": "fourth",
        },
      ]
    }
  />
</ThemeProvider>
`;
